# 01 Importing Libraries

In [2]:
# import libraries
import pandas as pd
import numpy as np
import os

In [4]:
path = r'/Users/Dena/Instacart Basket Analysis'

In [5]:
# import product data
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'))

df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [6]:
# import order data
df_ords = pd.read_csv(r'/Users/Dena/Instacart Basket Analysis/02 Data/Prepared Data/orders_wrangled.csv', index_col = False)

df_ords.head()

Unnamed: 0  order_id  user_id  number_of_orders_by_user  \
0           0   2539329        1                         1   
1           1   2398795        1                         2   
2           2    473747        1                         3   
3           3   2254736        1                         4   
4           4    431534        1                         5   

   orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                   2                  8                     NaN  
1                   3                  7                    15.0  
2                   3                 12                    21.0  
3                   4                  7                    29.0  
4                   4                 15                    28.0

In [7]:
# drop the first unnecessary column
df_ords.drop(columns = ['Unnamed: 0'])

order_id  user_id  number_of_orders_by_user  orders_day_of_week  \
0         2539329        1                         1                   2   
1         2398795        1                         2                   3   
2          473747        1                         3                   3   
3         2254736        1                         4                   4   
4          431534        1                         5                   4   
...           ...      ...                       ...                 ...   
3421078   2266710   206209                        10                   5   
3421079   1854736   206209                        11                   4   
3421080    626363   206209                        12                   1   
3421081   2977660   206209                        13                   1   
3421082    272231   206209                        14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

In [8]:
# overwrite the previous data frame
df_ords = df_ords.drop(columns = ['Unnamed: 0'])

In [9]:
# check that the order data frame is looking good
df_ords.head()

order_id  user_id  number_of_orders_by_user  orders_day_of_week  \
0   2539329        1                         1                   2   
1   2398795        1                         2                   3   
2    473747        1                         3                   3   
3   2254736        1                         4                   4   
4    431534        1                         5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

# 02 Data Consistency Checks

In [10]:
df_ords.describe()

order_id       user_id  number_of_orders_by_user  \
count  3.421083e+06  3.421083e+06              3.421083e+06   
mean   1.710542e+06  1.029782e+05              1.715486e+01   
std    9.875817e+05  5.953372e+04              1.773316e+01   
min    1.000000e+00  1.000000e+00              1.000000e+00   
25%    8.552715e+05  5.139400e+04              5.000000e+00   
50%    1.710542e+06  1.026890e+05              1.100000e+01   
75%    2.565812e+06  1.543850e+05              2.300000e+01   
max    3.421083e+06  2.062090e+05              1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

In [11]:
# create a dataframe
df_test = pd.DataFrame()

In [12]:
# create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [13]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [14]:
# check for mixed types

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [15]:
# convert data type to string
df_test['mix'] = df_test['mix'].astype('str')

In [16]:
# find missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [17]:
# create a data frame of the missing values only
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [18]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [19]:
df_prods.shape

(49693, 5)

In [20]:
# create a new clean dataframe, just using the non-missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [21]:
df_prods_clean.shape

(49677, 5)

In [22]:
# check that 16 rows were deleted
49693 - 49677

16

In [23]:
# this the code to overwrite the dataframe with a new one, which is risky --
# df_prods[‘product_name’].dropna(inplace = True)

In [24]:
# find full duplicates in the data
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [25]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [26]:
df_prods_clean.shape

(49677, 5)

In [27]:
# drop the duplicates from the data frame
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()
df_prods_clean_no_dups.shape

(49672, 5)

In [28]:
# check that 5 rows were deleted
49677 - 49672

5

# 03 Export Product Data

In [29]:
# export new product data
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# 04 Task 4.5 Questions

### Question 2

In [30]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

#### There is something "off" in the prices column. The max value is 99999, which makes no sense.

### Questions 3 & 4

In [31]:
# check for mixed-type data

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

#### Nothing was printed, so there are no columns with mixed data types!

### Question 5

In [32]:
# check for missing values
df_ords.isnull().sum()

order_id                         0
user_id                          0
number_of_orders_by_user         0
orders_day_of_week               0
order_hour_of_day                0
days_since_prior_order      206209
dtype: int64

In [33]:
df_ords.shape

(3421083, 6)

In [34]:
df_ords[df_ords.days_since_prior_order == 0]

order_id  user_id  number_of_orders_by_user  orders_day_of_week  \
8         2295261        1                         9                   1   
43        2557754        4                         5                   5   
189       2148135       17                        30                   6   
296        965160       24                        19                   0   
337       3162630       27                        24                   2   
...           ...      ...                       ...                 ...   
3420980      4277   206206                        47                   2   
3420985    648225   206206                        52                   0   
3420991   2475308   206206                        58                   0   
3421001   1043943   206206                        68                   0   
3421044   2079142   206208                        26                   2   

         order_hour_of_day  days_since_prior_order  
8                       16                     0.0  
43                      13                     0.0  
189                     15                     0.0  
296                     16                     0.0  
337                     15                     0.0  
...                    ...                     ...  
3420980                 17                     0.0  
3420985                 19                     0.0  
3420991                 15                     0.0  
3421001                 20                     0.0  
3421044                 16                     0.0  

[67755 rows x 6 columns]

In [35]:
df_ords.head(40)

order_id  user_id  number_of_orders_by_user  orders_day_of_week  \
0    2539329        1                         1                   2   
1    2398795        1                         2                   3   
2     473747        1                         3                   3   
3    2254736        1                         4                   4   
4     431534        1                         5                   4   
5    3367565        1                         6                   2   
6     550135        1                         7                   1   
7    3108588        1                         8                   1   
8    2295261        1                         9                   1   
9    2550362        1                        10                   4   
10   1187899        1                        11                   4   
11   2168274        2                         1                   2   
12   1501582        2                         2                   5   
13   1901567        2                         3                   1   
14    738281        2                         4                   2   
15   1673511        2                         5                   3   
16   1199898        2                         6                   2   
17   3194192        2                         7                   2   
18    788338        2                         8                   1   
19   1718559        2                         9                   2   
20   1447487        2                        10                   1   
21   1402090        2                        11                   1   
22   3186735        2                        12                   1   
23   3268552        2                        13                   4   
24    839880        2                        14                   3   
25   1492625        2                        15                   1   
26   1374495        3                         1                   1   
27    444309        3                         2                   3   
28   3002854        3                         3                   3   
29   2037211        3                         4                   2   
30   2710558        3                         5                   0   
31   1972919        3                         6                   0   
32   1839752        3                         7                   0   
33   3225766        3                         8                   0   
34   3160850        3                         9                   0   
35    676467        3                        10                   3   
36    521107        3                        11                   0   
37   1402502        3                        12                   1   
38   2774568        3                        13                   5   
39   3343014        4                         1                   6   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
1                   7                    15.0  
2                  12                    21.0  
3                   7                    29.0  
4                  15                    28.0  
5                   7                    19.0  
6                   9                    20.0  
7                  14                    14.0  
8                  16                     0.0  
9                   8                    30.0  
10                  8                    14.0  
11                 11                     NaN  
12                 10                    10.0  
13                 10                     3.0  
14                 10                     8.0  
15                 11                     8.0  
16                  9                    13.0  
17                 12                    14.0  
18                 15                    27.0  
19                  9                     8.0  
20                 11                     6.0  
21                 10               

In [36]:
df_ords.tail(5)

order_id  user_id  number_of_orders_by_user  orders_day_of_week  \
3421078   2266710   206209                        10                   5   
3421079   1854736   206209                        11                   4   
3421080    626363   206209                        12                   1   
3421081   2977660   206209                        13                   1   
3421082    272231   206209                        14                   6   

         order_hour_of_day  days_since_prior_order  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0

#### There are 206,209 missing values in the "days_since_prior_order" column.
#### After exploring, I noticed a pattern that the first entry in this column for each new user is NaN.
#### There are 206,209 users, therefore my explanation is the very first order for each new user is listed as NaN.

### Question 6

In [37]:
# address the missing values using an appropriate method
df_ords.dtypes

order_id                      int64
user_id                       int64
number_of_orders_by_user      int64
orders_day_of_week            int64
order_hour_of_day             int64
days_since_prior_order      float64
dtype: object

In [38]:
# create a new column to flag if an order is the first order or not

df_ords['is_first_order'] = np.where(df_ords['days_since_prior_order'].isnull(),'Yes', 'No')
df_ords.head(40)

order_id  user_id  number_of_orders_by_user  orders_day_of_week  \
0    2539329        1                         1                   2   
1    2398795        1                         2                   3   
2     473747        1                         3                   3   
3    2254736        1                         4                   4   
4     431534        1                         5                   4   
5    3367565        1                         6                   2   
6     550135        1                         7                   1   
7    3108588        1                         8                   1   
8    2295261        1                         9                   1   
9    2550362        1                        10                   4   
10   1187899        1                        11                   4   
11   2168274        2                         1                   2   
12   1501582        2                         2                   5   
13   1901567        2                         3                   1   
14    738281        2                         4                   2   
15   1673511        2                         5                   3   
16   1199898        2                         6                   2   
17   3194192        2                         7                   2   
18    788338        2                         8                   1   
19   1718559        2                         9                   2   
20   1447487        2                        10                   1   
21   1402090        2                        11                   1   
22   3186735        2                        12                   1   
23   3268552        2                        13                   4   
24    839880        2                        14                   3   
25   1492625        2                        15                   1   
26   1374495        3                         1                   1   
27    444309        3                         2                   3   
28   3002854        3                         3                   3   
29   2037211        3                         4                   2   
30   2710558        3                         5                   0   
31   1972919        3                         6                   0   
32   1839752        3                         7                   0   
33   3225766        3                         8                   0   
34   3160850        3                         9                   0   
35    676467        3                        10                   3   
36    521107        3                        11                   0   
37   1402502        3                        12                   1   
38   2774568        3                        13                   5   
39   3343014        4                         1                   6   

    order_hour_of_day  days_since_prior_order is_first_order  
0                   8                     NaN            Yes  
1                   7                    15.0             No  
2                  12                    21.0             No  
3                   7                    29.0             No  
4                  15                    28.0             No  
5                   7                    19.0             No  
6                   9                    20.0             No  
7                  14                    14.0             No  
8                  16                     0.0             No  
9                   8                    30.0             No  
10                  8                    14.0             No  
11                 11                     NaN            Yes  
12                 10                    10.0             No  
13                 10                     3.0             No  
14                 10                     8.0             No  
15                 11                     8.0             No  
16                  9

In [39]:
df_ords['is_first_order'].value_counts(dropna = False)

No     3214874
Yes     206209
Name: is_first_order, dtype: int64

### Questions 7 & 8

In [40]:
# find full duplicate rows in the data
df_dups_2 = df_ords[df_ords.duplicated()]
df_dups_2

Empty DataFrame
Columns: [order_id, user_id, number_of_orders_by_user, orders_day_of_week, order_hour_of_day, days_since_prior_order, is_first_order]
Index: []

#### No rows displayed, therefore there are no duplicates in the data.

# 05 Export Order Data

In [41]:
# export new order data
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked_redo.csv'))